### PR TITLE
Add nix-channel-index (and also some minor refactoring)

### PR DESCRIPTION
**Pull request by K900**
This adds a new tool that allows generating SQLite databases compatible with the default Nix command-not-found implementation. The intent is to eventually replace the current command-not-found database generation code (which is painfully slow) with this implementation to hopefully massively speed up channel updates.

### CODE REVIEW

File: .envrc [hunk 1]
```diff
@@ -0,0 +1 @@
+use flake
```
Comment (bennofs) [inline .envrc:1]: I don't like putting .envrc in the repo. It means that anyone using direnv and wanting a different enviroment/not use the flake env has a hard time.

File: src/bin/nix-channel-index.rs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                if let Ok(binary) = path.strip_prefix("/bin") {
+                    let attr = origin.attr.clone();
+                    let system = origin.system.clone();
+                    let binary: String = binary.to_string_lossy().into();
+
+                    if !binary.starts_with('.') {
+                        connection
+                            .execute(
+                                "insert or replace into Programs(name, system, package) values (?, ?, ?)",
+                                (attr, system, binary),
```
Comment (bennofs) [inline src/bin/nix-channel-index.rs:175]: Is this correct? I don't know how the current programs sqlite db works but from the naming I would have expected "package" to be the attr and "name" the binary name?

File: src/nixpkgs.rs [hunk 3]
```diff
@@ -26,20 +26,24 @@ pub fn query_packages(
     nixpkgs: &str,
     scope: Option<&str>,
     show_trace: bool,
+    system: Option<&str>,
 ) -> PackagesQuery<ChildStdout> {
     let mut cmd = Command::new("nix-env");
     cmd.arg("-qaP")
         .arg("--out-path")
         .arg("--xml")
-        .arg("--arg")
```
Comment (bennofs) [inline src/nixpkgs.rs:34]: Why remove this?

**Review by bennofs — CHANGES_REQUESTED** [review]
Overall looks good to me. Since I merged some changes to master, this now has merge conflicts. Do you have time to rebase this to current master?

File: .envrc (hunk 1 above)
Comment (bennofs) [reply .envrc:1]: As an alternative, I suggest adding .envrc to gitignore so people can easily add a envrc themselves

File: src/nixpkgs.rs (hunk 3 above)
Comment (K900) [reply src/nixpkgs.rs:34]: Accident, will fix.

File: src/bin/nix-channel-index.rs (hunk 2 above)
Comment (K900) [reply src/bin/nix-channel-index.rs:175]: Good find. Will fix in a bit.

File: .envrc (hunk 1 above)
Comment (K900) [reply .envrc:1]: All right.

**Comment by K900** [comment]
With the current version of this PR, I can eval + index all of nixpkgs for the four major platforms ({aarch64,x86_64}-{linux,darwin}) in _under three minutes_.

**Comment by K900** [comment]
Hopefully last push of the night, fixed symlinks in `/bin` not being indexed.

**Comment by K900** [comment]
And of course just as I said that, I found another bug. We should ignore non-`toplevel` paths for c-n-f, as they're not installable.

**Comment by K900** [comment]
Hopefully the last few changes to get us closer to the old database (mostly based on carefully comparing things by eye). The delta for x86_64-linux  is down to ~2400 entries out of ~64000, and at a glance the new version mostly looks better.

**Comment by K900** [comment]
Database diff: https://gist.github.com/K900/5144715c4747da1871c16782ca0c083e

**Comment by ncfavier** [comment]
This has now [landed](https://github.com/NixOS/nixpkgs/pull/210054) in nixpkgs!

**Pull request merged**